presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftPageSource.java

```diff
@@ -0,0 +1,194 @@
+
+    private PrestoThriftId nextToken;
+    private boolean firstCall = true;
+    private CompletableFuture<PrestoThriftPage> future;
+    private long completedBytes;
+
+    public ThriftPageSource(
+            PrestoThriftServiceProvider clientProvider,
+            ThriftConnectorConfig config,
+            ThriftConnectorSplit split,
+            List<ColumnHandle> columns)
+    {
+        // init columns
+        requireNonNull(columns, "columns is null");
+        List<String> columnNames = new ArrayList<>(columns.size());
+        List<Type> columnTypes = new ArrayList<>(columns.size());
+        for (ColumnHandle columnHandle : columns) {
+            ThriftColumnHandle thriftColumnHandle = (ThriftColumnHandle) columnHandle;
+            columnNames.add(thriftColumnHandle.getColumnName());
+            columnTypes.add(thriftColumnHandle.getColumnType());
+        }
+        this.columnNames = unmodifiableList(columnNames);
+        this.columnTypes = unmodifiableList(columnTypes);
+
+        // init config variables
+        this.maxBytesPerResponse = requireNonNull(config, "config is null").getMaxResponseSize().toBytes();
+
+        // init split
+        requireNonNull(split, "split is null");
+        this.splitId = new PrestoThriftId(split.getSplitId());
+
+        // init client
+        requireNonNull(clientProvider, "clientProvider is null");
+        if (split.getAddresses().isEmpty()) {
+            this.client = clientProvider.anyHostClient();
+        }
+        else {
+            this.client = clientProvider.selectedHostClient(split.getAddresses());
+        }
+    }
+
+    @Override
+    public long getTotalBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return completedBytes;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return readTimeNanos.get();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return !firstCall && !canGetMoreData(nextToken);
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        if (future != null) {
+            if (!future.isDone()) {
+                // data request is in progress
+                return null;
+            }
+            else {
+                // response for data request is ready
+                PrestoThriftPage rowsBatch = getFutureValue(future);
+                Page result = processBatch(rowsBatch);
+                // immediately try sending a new request
+                if (canGetMoreData(nextToken)) {
+                    future = sendDataRequestInternal();
+                }
+                else {
+                    future = null;
+                }
+                return result;
+            }
+        }
+        else {
+            // no data request in progress
+            if (firstCall || canGetMoreData(nextToken)) {
```

Comment:
I think it should never happen in the current form, but let's keep it in case I missed anything. Previously it was necessary, because the request wasn't immediately sent.